gadgets/src/permutation.rs

```diff
@@ -0,0 +1,312 @@
+//! Chip that implements permutation fingerprints
+//! fingerprints &= \prod_{row_j} \left ( \alpha - \sum_k (\gamma^k \cdot cell_j(k))  \right )
+//! power of gamma are defined in columns to trade more columns with less degrees
+use std::iter;
+#[rustfmt::skip]
+// | q_row_non_first | q_row_enable | alpha     | gamma     | gamma power 2   | ... | row fingerprint | accmulated fingerprint |
+// |-----------------|--------------|-----------|-----------|-----------------|     | --------------- | ---------------------- |
+// | 0               |1             |alpha      | gamma     | gamma **2       | ... |  F              |  F                     |
+// | 1               |1             |alpha      | gamma     | gamma **2       | ... |  F              |  F                     |
+// | 1               |1             |alpha      | gamma     | gamma **2       | ... |  F              |  F                     |
+
+use std::marker::PhantomData;
+
+use crate::util::Expr;
+use eth_types::Field;
+use halo2_proofs::{
+    circuit::{AssignedCell, Region, Value},
+    plonk::{Advice, Column, ConstraintSystem, Error, Expression, Selector},
+    poly::Rotation,
+};
+use itertools::Itertools;
+
+/// Config for PermutationChipConfig
+#[derive(Clone, Debug)]
+pub struct PermutationChipConfig<F> {
+    // column
+    acc_fingerprints: Column<Advice>,
+    row_fingerprints: Column<Advice>,
+    alpha: Column<Advice>,
+    power_of_gamma: Vec<Column<Advice>>,
+    // selector
+    q_row_non_first: Selector, // 1 between (first, end], exclude first
+    q_row_enable: Selector,    // 1 for all rows (including first)
+
+    _phantom: PhantomData<F>,
+}
+
+type PermutationAssignedCells<F> = (
```

Comment:
Maybe it would be helpful to have a comment here to indicate the values of each cell. Something like
`alpha`, `gamma`, `previous fingerprint accumulator`, `last fingerprint`.

Reply:
documentation added